tests/src/test/java/alluxio/server/health/MetricMasterIntegrityIntegrationTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.server.health;
+
+import alluxio.master.LocalAlluxioCluster;
+import alluxio.master.metrics.MetricsMaster;
+import alluxio.metrics.Metric;
+import alluxio.testutils.LocalAlluxioClusterResource;
+
+import alluxio.util.executor.ExecutorServiceFactories;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+public class MetricMasterIntegrityIntegrationTest {
+  @Rule
+  public LocalAlluxioClusterResource mClusterResource =
+      new LocalAlluxioClusterResource.Builder().build();
+  private LocalAlluxioCluster mCluster;
+  private MetricsMaster mMetricsMaster;
+  private ExecutorService mExecutor;
+  private int mNumWorkers = 3000;
+
+  @Before
+  public void before() {
+    mCluster = mClusterResource.get();
+    mMetricsMaster = mCluster.getLocalAlluxioMaster().getMasterProcess().getMaster(MetricsMaster.class);
+    mExecutor = ExecutorServiceFactories.fixedThreadPool("alluxio-test", 50).create();
+  }
+
+  @Test
+  public void testSameWorkerManyHeartbeats() throws Exception {
+    Map<String, List<Metric>> map = mMetricsMaster.getWorkerMetrics();
+    String hostname = "10.0.0.240";
+    List<Metric> metrics = map.get(hostname);
+    List<Future<?>> futureList = new ArrayList<>();
+    long start = System.currentTimeMillis();
+    for (int i = 0; i < 200; i++) {
+      Future<?> future = mExecutor.submit(() -> mMetricsMaster.putWorkerMetric(hostname, metrics));
+      futureList.add(future);
+    }
+    long mid = System.currentTimeMillis();
+    System.out.println("submit task takes " + (mid - start));
+    for (Future<?> future : futureList) {
+      future.get();
+    }
+    System.out.println("get result takes " + (System.currentTimeMillis() - mid));
+  }
+
+  @Test
+  public void testMultipleWorkersOneHeartbeat() throws Exception {
```

Comment:
@gpang For this test, I test with numWorkers = 50/200/1000/3000 cases and executor thread= 10/50.
For all the tests, in master branch, the throughput is 1000-3000 heartbeats/second. After my changes in https://github.com/Alluxio/alluxio/pull/10493, the throughput is 2300 to 5100 heartbeats/second. The throughput nearly doubled. 

I cannot get the real circumstances we faced before.  one of my guesses is that the bottleneck is memory other than the synchronization lock. @ZacBlanco Mentioned before that the `Metric.getFullMetricName` do many string concatenation and cause many memory copies. This function may use a large amount memory in reality.

Reply:
```
  /**
   * @return the fully qualified metric name, which is of pattern
   *         instance.[hostname-id:instanceId.]name[.tagName:tagValue]*, where the tags are appended
   *         at the end
   */
  public String getFullMetricName() {
    StringBuilder sb = new StringBuilder();
    sb.append(mInstanceType).append('.');
    if (mHostname != null) {
      sb.append(mHostname);
      if (mInstanceId != null) {
        sb.append(ID_SEPARATOR).append(mInstanceId);
      }
      sb.append('.');
    }

    sb.append(mName);
    for (Entry<String, String> entry : mTags.entrySet()) {
      sb.append('.').append(entry.getKey()).append(TAG_SEPARATOR).append(entry.getValue());
    }
    return sb.toString();
  }
```